File: examples/kafka-consumer-python/python/src/kafka_consumer_demo/consumer_main.py
"""Optional real-consumer entrypoint (requires running Kafka)."""

from __future__ import annotations

import os
import sys

from kafka import KafkaConsumer

from kafka_consumer_demo.processor import DecodedMessage, process_poll_result


def _print_message(m: DecodedMessage) -> None:
    print(f"topic={m.topic} key={m.key_text!r} value={m.value_text!r}", flush=True)


def main() -> int:
    bootstrap = os.environ.get("KAFKA_BOOTSTRAP_SERVERS", "localhost:9092")
    topic = os.environ.get("KAFKA_TOPIC", "demo.events")
    group = os.environ.get("KAFKA_GROUP_ID", "pykotmig-kafka-demo")
    consumer = KafkaConsumer(
        topic,
        bootstrap_servers=bootstrap,
        group_id=group,
        auto_offset_reset="earliest",
        enable_auto_commit=True,
        consumer_timeout_ms=5000,
    )
    try:
        while True:
            raw = consumer.poll(timeout_ms=1000)
            if not raw:
                continue
            process_poll_result(raw, _print_message)
    except KeyboardInterrupt:
        print("stopped", file=sys.stderr)
    finally:
        consumer.close()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/kafka-consumer-python/kotlin/bin/main/dev/pykotmig/kafkaconsumer/ConsumerMain.kt
package dev.pykotmig.kafkaconsumer

import org.apache.kafka.clients.consumer.ConsumerConfig
import org.apache.kafka.clients.consumer.KafkaConsumer
import org.apache.kafka.common.serialization.ByteArrayDeserializer
import java.time.Duration
import java.util.Properties

private fun env(name: String, default: String): String = System.getenv(name)?.takeIf { it.isNotBlank() } ?: default

fun main() {
    val bootstrap = env("KAFKA_BOOTSTRAP_SERVERS", "localhost:9092")
    val topic = env("KAFKA_TOPIC", "demo.events")
    val group = env("KAFKA_GROUP_ID", "pykotmig-kafka-demo")
    val props =
        Properties().apply {
            put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrap)
            put(ConsumerConfig.GROUP_ID_CONFIG, group)
            put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest")
            put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "true")
            put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer::class.java.name)
            put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer::class.java.name)
            put(ConsumerConfig.CONNECTIONS_MAX_IDLE_MS_CONFIG, "10000")
        }

    KafkaConsumer<ByteArray?, ByteArray?>(props).use { consumer ->
        consumer.subscribe(listOf(topic))
        while (true) {
            val batch = consumer.poll(Duration.ofMillis(1000))
            if (batch.isEmpty) {
                continue
            }
            processPollResult(batch.toPollResult()) { m ->
                println("topic=${m.topic} key=${m.keyText} value=${m.valueText}")
            }
        }
    }
}
